R/dependecies.R

```diff
@@ -0,0 +1,437 @@
+  names[[1]] <- names[[1]][!names[[1]] %in% core_pkgs]
+  list2 <- list2[!names[[2]] %in% core_pkgs]
+  names[[2]] <- names[[2]][!names[[2]] %in% core_pkgs]
+
+  if (length(list1) == 0) return(list2)
+  if (length(list2) == 0) return(list1)
+
+  names(list1) <- names[[1]]
+  names(list2) <- names[[2]]
+
+  ## Find the rightmost dependency of the current package and move our current
+  ## package to the immediate right of that spot.  This preserves previous
+  ## sorting order while ensuring that the current package will be installed
+  ## after all its dependencies
+  if (current_parent %in% names[[2]] && any(names[[2]] %in% names[[1]])) {
+    init_parent_slot <- which(names[[2]] == current_parent)
+    final_parent_slot <- max(which(names[[2]] %in% names[[1]]))
+    if (final_parent_slot > init_parent_slot) {
+      sel1 <- seq_along(names[[2]]) != init_parent_slot & seq_along(names[[2]]) <= final_parent_slot
+      sel2 <- seq_along(names[[2]]) > final_parent_slot
+      list2 <- c(list2[sel1], list2[init_parent_slot], list2[sel2])
+      names[[2]] <- c(names[[2]][sel1], names[[2]][init_parent_slot], names[[2]][sel2])
+    }
+  }
+
+  list2 <- swap_packages(names[[1]], names[[2]], list1, list2)
+
+  c(list1[!names[[1]] %in% names[[2]]], list2)
+}
+
+#' Swap packages by comparing version information. 1ist2 has already
+#' been ordered by dependency, so is imperative that we keep its order while
+#' potentially swapping in the corresponding packages in list1 that have later
+#' version requirements.
+swap_packages <- function(names1, names2, list1, list2) {
+  ## Swap packages when the package in list1 is more recent (> version) than
+  ## it's corresponding package in list2.  Keep the list 2 element if it is a
+  ## locked package, as well.  If their remotes are not identical and neither
+  ## has specified a version, then keep whichever remote has a more recent version
+  swap_package2for1 <- vapply(
+    names1
+    , function(n) {
+      if (!n %in% names2) return(FALSE)
+      obj1 <- list1[[n]]
+      obj2 <- list2[[n]]
+      if (obj1$is_dependency_package && obj2$is_dependency_package) {
+        if (is.na(obj1$version) && is.na(obj2$version)) {
+          (obj1$remote != obj2$remote) &&
+            package_version(obj1$latest_version) >
+            package_version(obj2$latest_version)
+        } else {
+          !is.na(obj1$version) && (is.na(obj2$version) ||
+            package_version(obj1$version) > package_version(obj2$version))
+        }
+      } else {
+        obj2$is_dependency_package
+      }
+    }
+    , logical(1))
+
+  ## Delete temporary files for swapped items if they are distinct
+  lapply(which(swap_package2for1)
+    , function(index) {
+      obj2 <- list2[which(names2 == names1[index])]
+      obj1 <- list1[[index]]
+      if (!is.null(obj1$download_path) && !is.null(obj2$download_path) &&
+        obj2$download_path != obj1$download_path) {
+        unlink(obj2$download_path, TRUE, TRUE)
```

Comment:
Shouldn't the directory with downloaded files be unlinked anyway after the `lockbox` call? This may be redundant.